beacon-chain/sync/gossipsub_controller.go

```diff
@@ -0,0 +1,213 @@
+	}
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	// register topic families for the current NSE -> this is idempotent
+	for _, family := range families {
+		key := topicFamilyKeyFrom(family)
+		existing, seen := g.activeTopicFamilies[key]
+		if !seen {
+			g.activeTopicFamilies[key] = family
+			existing = family
+			log.WithFields(logrus.Fields{
+				"topicName":  key.topicName,
+				"forkDigest": fmt.Sprintf("%#x", key.forkDigest),
+				"epoch":      currentEpoch,
+			}).Info("Registered topic family")
+		}
+
+		switch tf := existing.(type) {
+		case GossipsubTopicFamilyWithDynamicSubnets:
+			tf.UnsubscribeForSlot(slot)
+			tf.SubscribeForSlot(slot)
+		case GossipsubTopicFamilyWithoutDynamicSubnets:
+			if !seen {
+				tf.Subscribe()
+			}
+		}
+	}
+
+	// remove topic families for the previous NSE -> this is idempotent
+	if beyond, previous := isOneEpochBeyondForkBoundary(currentEpoch); beyond {
+		for key, family := range g.activeTopicFamilies {
+			if key.forkDigest == previous.ForkDigest {
+
+				family.UnsubscribeAll()
+				delete(g.activeTopicFamilies, key)
+
+				log.WithFields(logrus.Fields{
+					"topicName":  key.topicName,
+					"forkDigest": fmt.Sprintf("%#x", key.forkDigest),
+				}).Info("Removed topic family")
+			}
+		}
+	}
+}
+
+func (g *GossipsubController) Stop() {
+	g.cancel()
+	g.wg.Wait()
+
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	for _, family := range g.activeTopicFamilies {
+		family.UnsubscribeAll()
+	}
+}
+
+func (g *GossipsubController) GetCurrentActiveTopics() []string {
+	g.mu.RLock()
+	defer g.mu.RUnlock()
+
+	slot := g.syncService.cfg.clock.CurrentSlot()
+	var topics []string
+	for _, f := range g.activeTopicFamilies {
+		tfm, ok := f.(GossipsubTopicFamilyWithDynamicSubnets)
+		if !ok {
+			continue
+		}
+		topics = append(topics, tfm.TopicsToSubscribeForSlot(slot)...)
+	}
+	return topics
+}
+
+func (g *GossipsubController) ExtractTopics(_ context.Context, node *enode.Node) ([]string, error) {
+	if node == nil {
+		return nil, errors.New("enode is nil")
+	}
+
+	g.mu.RLock()
+	defer g.mu.RUnlock()
+
+	families := make([]GossipsubTopicFamilyWithDynamicSubnets, 0, len(g.activeTopicFamilies))
+	for _, f := range g.activeTopicFamilies {
+		if tfm, ok := f.(GossipsubTopicFamilyWithDynamicSubnets); ok {
+			families = append(families, tfm)
+		}
+	}
+
+	// Collect topics from dynamic families only, de-duplicated.
+	topicSet := make(map[string]struct{})
+	for _, df := range families {
+		topics, err := df.ExtractTopicsForNode(node)
+		if err != nil {
+			log.WithError(err).WithFields(logrus.Fields{
+				"topicFamily": fmt.Sprintf("%T", df),
+			}).Debug("Failed to get topics for node from family")
+			continue
+		}
+		for _, t := range topics {
+			topicSet[t] = struct{}{}
+		}
+	}
+
+	out := make([]string, 0, len(topicSet))
+	for t := range topicSet {
+		out = append(out, t)
+	}
+	return out, nil
+}
+
+func isNextEpochForkBoundary(currentEpoch primitives.Epoch) (bool, params.NetworkScheduleEntry) {
```

Comment:
The interface of these funcs is a little awkward, and I also noticed that in the only place they are used, you already have the current epoch NSE. My suggestion is to just use the NSEs directly to compare them. Also there is a helper to get the next entry: `params.NextNetworkScheduleEntry`. You could add `params.PreviousNetworkScheduleEntry`.

Reply:
have improved this and introduced a "params.PreviousNetworkScheduleEntry" function.